truelearn/learning/_ink_classifier.py

```diff
@@ -1,25 +1,285 @@
+            bias_weight=bias_weight,
+        )
+
+        if novelty_classifier is None:
+            novelty_classifier = NoveltyClassifier()
+        if interest_classifier is None:
+            interest_classifier = InterestClassifier()
+
+        self.novelty_classifier = novelty_classifier
+        self.interest_classifier = interest_classifier
+        self.tau = tau
+        self.threshold = threshold
+        self.greedy = greedy
+
+        if novelty_weight is None:
+            novelty_weight = {"mean": 0.0, "variance": 1.0}
+        else:
+            if "mean" not in novelty_weight or "variance" not in novelty_weight:
+                raise ValueError(
+                    f'The novelty_weight must contain "mean" and "variance".'
+                    f" Got {novelty_weight} instead."
+                )
+
+        if interest_weight is None:
+            interest_weight = {"mean": 0.0, "variance": 1.0}
+        else:
+            if "mean" not in interest_weight or "variance" not in interest_weight:
+                raise ValueError(
+                    f'The interest_weight must contain "mean" and "variance".'
+                    f" Got {interest_weight} instead."
+                )
+
+        if bias_weight is None:
+            bias_weight = {"mean": 0.0, "variance": 1.0}
+        else:
+            if "mean" not in bias_weight or "variance" not in bias_weight:
+                raise ValueError(
+                    f'The bias_weight must contain "mean" and "variance".'
+                    f" Got {bias_weight} instead."
+                )
+
+        self.novelty_weight: Dict[str, float] = novelty_weight
+        self.interest_weight: Dict[str, float] = interest_weight
+        self.bias_weight: Dict[str, float] = bias_weight
+
+        self.__env = trueskill.setup(
+            mu=0.0,
+            sigma=INKClassifier.DEFAULT_SIGMA,
+            beta=1,
+            tau=tau,
+            draw_probability=INKClassifier.DEFAULT_DRAW_PROBA,
+            backend="mpmath",
+        )
+
+    def __calculate_sum_prediction(
+        self,
+        *,
+        mu_novelty: float,
+        var_novelty: float,
+        pred_novelty: float,
+        mu_interest: float,
+        var_interest: float,
+        pred_interest: float,
+        mu_bias: float,
+        var_bias: float,
+        pred_bias: float,
+    ) -> float:
+        difference = (
+            (mu_novelty * pred_novelty)
+            + (mu_interest * pred_interest)
+            + (mu_bias * pred_bias)
+            - self.threshold
+        )
+        std = math.sqrt(
+            ((var_novelty) * pred_novelty)
+            + ((var_interest) * pred_interest)
+            + ((var_bias) * pred_bias)
+        )
+        return statistics.NormalDist(mu=0, sigma=std).cdf(difference)
+
+    def __update_weights(
+        self, pred_novelty: float, pred_interest: float, pred_actual: float
+    ) -> None:
+        mu_novelty, var_novelty = (
```

Comment:
duplicate code warning from 163-174 and 246-257

Reply:
I think this is probably unavoidable, since I'm just retrieving data from each weight. I could write a function, but that would require me to return a 6-element tuple, which is much weirder than this.

Reply:
> I think this is probably unavoidable, since I'm just retrieving data from each weight. I could write a function, but that would require me to return a 6-element tuple, which is much weirder than this.

Maybe wrap the weights in some data class and then return that?

Reply:
I think it's a bit too much for some simple data retrievals. Maintaining such an inner class in our "INKClassifier" would be quite dirty.

Reply:
> I think it's a bit too much for some simple data retrievals. Maintaining such an inner class in our "INKClassifier" would be quite dirty.

Why dont you instead just call predict proba in update weights, it seems like there is no difference, so you get the sum prediction from predict proba and then use it in update weights

Reply:
> > I think it's a bit too much for some simple data retrievals. Maintaining such an inner class in our "INKClassifier" would be quite dirty.
> 
> Why dont you instead just call predict proba in update weights, it seems like there is no difference, so you get the sum prediction from predict proba and then use it in update weights nvm I see a difference

maybe a method then instead for get_current_pred containing all those lines returning the value and predict prob and update weights could use it

Reply:
> > I think it's a bit too much for some simple data retrievals. Maintaining such an inner class in our "INKClassifier" would be quite dirty.
> 
> Why dont you instead just call predict proba in update weights, it seems like there is no difference, so you get the sum prediction from predict proba and then use it in update weights nvm I see a difference

It seems that I fix this now. Please see my latest commit later.

Reply:
I've added a new commit to fix this.

Reply:
> > > I think it's a bit too much for some simple data retrievals. Maintaining such an inner class in our "INKClassifier" would be quite dirty.
> > 
> > 
> > Why dont you instead just call predict proba in update weights, it seems like there is no difference, so you get the sum prediction from predict proba and then use it in update weights nvm I see a difference
> 
> maybe a method then instead for get_current_pred containing all those lines returning the value and predict prob and update weights could use it. Essentially the same as predict_proba but also taking in params for pred_novelty and pred_interest

Seems like you sorted it out, I was getting confused 👍